databricks_cli/stack/api.py

```diff
@@ -139,6 +162,84 @@ def deploy_config(self, stack_config, stack_status=None, **kwargs):
+            click.echo(
+                "Downloading workspace asset '{}' with properties \n{}"
+                .format(
+                    resource_id, json.dumps(resource_properties, indent=2, separators=(',', ': '))
+                )
+            )
+            overwrite = kwargs.get('overwrite_notebooks', False)
+            self._download_workspace(resource_properties, overwrite)
+        else:
+            click.echo("Resource service '{}' not supported for download. "
+                       "skipping.".format(resource_service))
+
+    def _download_workspace(self, resource_properties, overwrite):
+        """
+        Deploy workspace asset.
+        TODO (alinxie) Change name to overwrite to be more toward overwriting exclusive resources.
+        For a notebook workspace resource, it is recommended to provide 'language' and 'format'
+        :param resource_properties: dict of properties for the workspace asset. Must contain the
+        'source_path' and 'path' fields. The other fields will be inferred if not provided.
+        :param overwrite: Whether or not to overwrite the contents of workspace notebooks.
+        :return: (dict, dict) of (physical_id, deploy_output). physical_id is the physical ID for
+        the stack status that contains the workspace path of the notebook or directory on datbricks.
+        deploy_output is the initial information about the asset on databricks at deploy time
+        returned by the REST API.
+        """
+        # Required fields. TODO(alinxie) put in _validate_config
+        local_path = resource_properties.get('source_path')
+        workspace_path = resource_properties.get('path')
+        object_type = resource_properties.get('object_type')
+        click.echo('Downloading {} from Databricks path {} to {}'.format(object_type,
+                                                                         workspace_path,
+                                                                         local_path))
+        if object_type == 'NOTEBOOK':
+            # Inference of notebook language and format. A tuple of (language, fmt) or Nonetype.
+            language_fmt = WorkspaceLanguage.to_language_and_format(local_path)
+            if language_fmt is None:
+                raise StackError("Workspace Notebook language and format cannot be inferred."
+                                 "Please check file extension of notebook 'source_path'.")
+            fmt = language_fmt[1]
+            local_dir = os.path.dirname(os.path.abspath(local_path))
+            if not os.path.exists(local_dir):
+                os.makedirs(local_dir)
+            self.workspace_client.export_workspace(workspace_path, local_path, fmt, overwrite)
+        elif object_type == 'DIRECTORY':
+            if not os.path.exists(local_path):
+                os.makedirs(local_path)
+            self.workspace_client.export_workspace_dir(workspace_path, local_path, overwrite)
```

Comment:
What is the behavior of this functions on folders?
(1) if overwrite, then is it a per-file overwrite. In other words, is the whole folder going to be erased first?
(2) if not overwrite, what is the behavior if (a) the folder exists but is empty (b) the folder exists and have files with different names.

Reply:
Seems like it is per-file overwrite, but I'll look into it and the cases you described.